clicommand/agent_start.go

```diff
@@ -1392,8 +1396,26 @@ func handlePoolSignals(ctx context.Context, l logger.Logger, pool *agent.AgentPo
 					l.Info("Received CTRL-C, send again to forcefully kill the agent(s)")
 					pool.Stop(true)
 				} else {
-					l.Info("Forcefully stopping running jobs and stopping the agent(s)")
-					pool.Stop(false)
+					l.Info("Forcefully stopping running jobs and stopping the agent(s) in %d seconds", cancelGracePeriod)
+
+					gracefulContext, _ := job.WithGracePeriod(ctx, cancelGracePeriodDuration(cancelGracePeriod)*time.Second)
+
+					go func() {
+						l.Info("Forced agent(s) to stop")
+						pool.Stop(false) // one last chance to stop
+
+						time.Sleep(cancelGracePeriodDuration(cancelGracePeriod/2) * time.Second)
+
+						l.Info("Cancelling all internal tasks and API requests")
+						cancel() // cancel the context to stop all network operations
+					}()
+
+					<-gracefulContext.Done()
```

Comment:
This section is unusual since it's what happens if the program hasn't already exited, but looks like part of the "normal flow" once the second interrupt is received. May be worth a comment explaining why "forceful stop" within the grace period doesn't lead to `os.Exit(1)`: 

```suggestion
					// Once pending retries and requests are cancelled, 
					// the main goroutine should exit before this grace period expires, ending the program.
					// If that doesn't happen, exit 1 below.
					<-gracefulContext.Done()
```